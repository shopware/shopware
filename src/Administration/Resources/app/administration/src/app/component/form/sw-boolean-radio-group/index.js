import './sw-boolean-radio-group.scss';

const { Component } = Shopware;

/**
 * @package admin
 *
 * @deprecated tag:v6.6.0 - Will be private
 * @public
 * @description radio input field for boolean and named entries.
 * @status ready
 * @example-type static
 * @component-example
 * <sw-boolean-radio-group
 *      label="Bool Radio group example"
 *      labelOptionTrue="Gross"
 *      labelOptionFalse="Net"
 *      :bordered="bordered">
 * </sw-boolean-radio-group>
 */
Component.register('sw-boolean-radio-group', {

    template:
`
<sw-radio-field
    class="sw-boolean-radio-group"
    v-bind="$attrs"
    :options="options"
    v-model="castedValue"
    :bordered="bordered">
</sw-radio-field>
`,
    model: {
        prop: 'value',
        event: 'change',
    },

    props: {
        value: {
            type: Boolean,
            required: false,
            // TODO: Boolean props should only be opt in and therefore default to false
            // eslint-disable-next-line vue/no-boolean-default
            default: true,
        },

        labelOptionTrue: {
            type: String,
            required: true,
        },

        labelOptionFalse: {
            type: String,
            required: true,
        },

        bordered: {
            type: Boolean,
            required: false,
            default: false,
        },
    },

    computed: {
        options() {
            return [
                { value: 'true', name: this.labelOptionTrue },
                { value: 'false', name: this.labelOptionFalse },
            ];
        },

        castedValue: {
            get() {
                return this.value ? this.value.toString() : 'false';
            },

            set(val) {
                this.$emit('change', val === 'true');
            },
        },
    },
});
