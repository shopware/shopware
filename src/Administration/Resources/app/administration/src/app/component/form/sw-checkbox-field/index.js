import template from './sw-checkbox-field.html.twig';
import './sw-checkbox-field.scss';

const { Component, Mixin } = Shopware;
const utils = Shopware.Utils;

/**
 * @package admin
 *
 * @deprecated tag:v6.6.0 - Will be private
 * @public
 * @description Boolean input field based on checkbox.
 * @status ready
 * @example-type static
 * @component-example
 * <sw-checkbox-field v-model="aBooleanProperty" label="Name"></sw-checkbox-field>
 */
Component.register('sw-checkbox-field', {
    template,
    inheritAttrs: false,

    mixins: [
        Mixin.getByName('sw-form-field'),
        Mixin.getByName('remove-api-error'),
    ],

    model: {
        prop: 'value',
        event: 'change',
    },

    props: {
        disabled: {
            type: Boolean,
            required: false,
            default: false,
        },

        value: {
            type: Boolean,
            required: false,
            // TODO: Boolean props should only be opt in and therefore default to false
            // eslint-disable-next-line vue/no-boolean-default
            default: null,
        },

        inheritedValue: {
            type: Boolean,
            required: false,
            // TODO: Boolean props should only be opt in and therefore default to false
            // eslint-disable-next-line vue/no-boolean-default
            default: null,
        },

        ghostValue: {
            type: Boolean,
            required: false,
            // TODO: Boolean props should only be opt in and therefore default to false
            // eslint-disable-next-line vue/no-boolean-default
            default: null,
        },

        error: {
            type: Object,
            required: false,
            default: null,
        },

        bordered: {
            type: Boolean,
            required: false,
            default: false,
        },
    },

    data() {
        return {
            currentValue: this.value,
            id: utils.createId(),
        };
    },

    computed: {
        swCheckboxFieldClasses() {
            return {
                'has--error': this.hasError,
                'is--disabled': this.disabled,
                'is--inherited': this.isInherited,
                'sw-field__checkbox--ghost': this.ghostValue,
                'is--bordered': this.bordered,
            };
        },

        identification() {
            return `sw-field--${this.id}`;
        },

        hasError() {
            return this.error && this.error.code !== 0;
        },

        inputState() {
            return this.currentValue || false;
        },

        isInheritanceField() {
            if (this.$attrs.isInheritanceField) {
                return true;
            }
            return this.inheritedValue !== null;
        },

        isInherited() {
            if (this.$attrs.isInherited) {
                return true;
            }
            return this.isInheritanceField && this.currentValue === null;
        },
    },

    watch: {
        value() { this.currentValue = this.value; },
    },

    methods: {
        onChange(changeEvent) {
            this.$emit('change', changeEvent.target.checked);
        },
    },
});
