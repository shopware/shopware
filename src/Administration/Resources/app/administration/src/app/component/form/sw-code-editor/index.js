import Ace from 'ace-builds/src-noconflict/ace';
import { setCompleters, textCompleter } from 'ace-builds/src-noconflict/ext-language_tools';
import 'ace-builds/src-noconflict/mode-twig';
import template from './sw-code-editor.html.twig';
import './sw-code-editor.scss';

const { Component } = Shopware;
const utils = Shopware.Utils;

/**
 * @public
 * @status ready
 * @description
 * Renders a code editor
 * @example-type dynamic
 * @component-example
 * <sw-code-editor label="Description">
 * </sw-code-editor>
 */
Component.register('sw-code-editor', {
    template,

    inject: [
        'feature',
        'userInputSanitizeService',
    ],

    props: {
        value: {
            type: String,
            required: false,
            default: '',
        },

        label: {
            type: String,
            required: false,
            default: '',
        },

        completerFunction: {
            type: Function,
            required: false,
            default: null,
        },

        editorConfig: {
            type: Object,
            required: false,
            default() {
                return {};
            },
        },

        completionMode: {
            type: String,
            required: false,
            default: 'text',
            validValues: ['entity', 'text'],
            validator(value) {
                if (!value.length) {
                    return true;
                }
                return ['entity', 'text'].includes(value);
            },
        },

        mode: {
            type: String,
            required: false,
            default: 'twig',
            validValues: ['twig', 'text'],
            validator(value) {
                if (!value.length) {
                    return true;
                }
                return ['twig', 'text'].includes(value);
            },
        },

        softWraps: {
            type: Boolean,
            required: false,
            default: true,
        },

        // set focus to the component when initially mounted
        setFocus: {
            type: Boolean,
            required: false,
            default: false,
        },

        disabled: {
            type: Boolean,
            required: false,
            default: false,
        },

        sanitizeInput: {
            type: Boolean,
            required: false,
            default: false,
        },

        sanitizeFieldName: {
            type: String,
            required: false,
            default: null,
        },
    },

    data() {
        return {
            editor: {},
            editorId: utils.createId(),
            contentWasSanitized: false,
        };
    },

    computed: {
        aceConfig() {
            return {
                ...{
                    mode: `ace/mode/${this.mode}`,
                    showPrintMargin: false,
                    wrap: this.softWraps,
                    readOnly: this.disabled,
                },
                ...this.editorConfig,
            };
        },
    },

    watch: {
        value(value) {
            if (value !== null && value !== this.editor.getValue()) {
                this.editor.setValue(value, 1);
            }
        },

        softWraps() {
            this.editor.session.setOption('wrap', this.softWraps);
        },
    },

    mounted() {
        this.mountedComponent();
    },

    destroyed() {
        this.destroyedComponent();
    },

    methods: {
        mountedComponent() {
            this.editor = Ace.edit(this.$refs['editor'.concat(this.editorId)], this.aceConfig);

            this.defineAutocompletion(this.completerFunction);

            this.editor.setValue(this.value || '', 1);
            this.editor.on('input', this.onInput);
            this.editor.on('blur', this.onBlur);

            if (this.setFocus) {
                this.editor.focus();
            }
        },

        destroyedComponent() {
            delete this.editor;
        },

        onInput() {
            const value = this.editor.getValue();

            if (this.value !== value) {
                this.$emit('input', value);
            }
        },

        async onBlur() {
            const value = await this.sanitizeEditorInput(this.editor.getValue());

            this.$emit('blur', value);
        },

        async sanitizeEditorInput(value) {
            if (
                this.feature.isActive('FEATURE_NEXT_15172') &&
                this.sanitizeInput
            ) {
                this.contentWasSanitized = false;

                if (this.value !== value) {
                    try {
                        const sanitizedValue = await this.userInputSanitizeService.sanitizeInput({
                            html: value,
                            field: this.sanitizeFieldName,
                        });

                        this.contentWasSanitized = sanitizedValue?.preview !== value;

                        if (this.contentWasSanitized) {
                            this.editor.setValue(sanitizedValue?.preview ?? value, 1);
                            return this.editor.getValue();
                        }
                    } catch (ignore) { /* api endpoint did not work, keep user entry */ }
                }
            }
            return value;
        },

        defineAutocompletion(completerFunction) {
            /**
             * Sets a completer function. If completitionMode is set to "entity"
             * Autocomplete can handle [x] and . otherwise it uses the default
             * RegEx from ACE.
             * The ExecCommands sets a hook to the "insertstring" command to
             * prevent the autocompletion-popup to hide after a .
             */
            setCompleters([]);
            if (completerFunction) {
                const textCompleterClonedPlain = JSON.parse(JSON.stringify(textCompleter));
                const textCompleterCloned = JSON.parse(JSON.stringify(textCompleter));

                if (this.completionMode === 'entity') {
                    textCompleterCloned.identifierRegexps = [/[\[\]\.a-zA-Z_0-9\$\-\u00A2-\uFFFF]/];

                    textCompleterCloned.getCompletions = function getComps(editor, session, pos, prefix, callback) {
                        this.identifierRegexps = [/[\[\][a-zA-Z_0-9\$\-\u00A2-\uFFFF]/];
                        callback(null, completerFunction(prefix));
                        this.identifierRegexps = [/[\[\]\.a-zA-Z_0-9\$\-\u00A2-\uFFFF]/];
                    };

                    textCompleterCloned.completerFunction = completerFunction;
                    this.editor.completers = [textCompleterCloned];

                    const startCallback = (function startCall(e) {
                        if (e.command.name === 'insertstring') {
                            if (e.args !== '\n' && e.args !== ' ') {
                                e.editor.execCommand('startAutocomplete', null);
                            }
                        }
                    });

                    this.editor.commands.on('afterExec', startCallback);
                } else {
                    textCompleterClonedPlain.identifierRegexps = null;
                    textCompleterClonedPlain.getCompletions = function getComps(editor, session, pos, prefix, callback) {
                        callback(null, completerFunction(prefix));
                    };

                    textCompleterClonedPlain.completerFunction = completerFunction;
                    this.editor.completers = [textCompleterClonedPlain];
                }
            } else {
                this.editor.completers = [];
            }
        },
    },
});
