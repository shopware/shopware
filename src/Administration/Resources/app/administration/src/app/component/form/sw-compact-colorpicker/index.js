import template from './sw-compact-colorpicker.html.twig';
import './sw-compact-colorpicker.scss';

const { Component } = Shopware;

/**
 * @package admin
 *
 * @private
 */
Component.extend('sw-compact-colorpicker', 'sw-colorpicker', {
    template,

    computed: {
        colorValue: {
            get() {
                return this.localValue;
            },
            set(newColor) {
                this.localValue = newColor;
            },
        },
    },

    methods: {
        emitColor() {
            this.$emit('input', this.localValue);
            this.visible = false;
        },
    },
});
