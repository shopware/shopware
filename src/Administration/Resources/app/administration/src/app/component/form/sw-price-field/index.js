import template from './sw-price-field.html.twig';
import './sw-price-field.scss';

const { Component, Application } = Shopware;
const utils = Shopware.Utils;

/**
 * @public
 * @status ready
 * @example-type static
 * @component-example
 * <sw-price-field :taxRate="{ taxRate: 19 }"
 *                 :price="[{ net: 10, gross: 11.90, currencyId: '...' }, ...]"
 *                 :defaultPrice="{...}"
 *                 :currency="{...}">
 * </sw-price-field>
 */
Component.register('sw-price-field', {
    template,

    inheritAttrs: false,

    props: {
        price: {
            type: Array,
            required: true,
            default() {
                return [];
            }
        },

        defaultPrice: {
            type: Object,
            required: false,
            default() {
                return {};
            }
        },

        taxRate: {
            type: Object,
            required: true,
            default() {
                return {};
            }
        },

        currency: {
            type: Object,
            required: true,
            default() {
                return {};
            }
        },

        validation: {
            required: false,
            default: null
        },

        label: {
            required: false,
            default: true
        },

        compact: {
            required: false,
            default: false
        },

        error: {
            type: Object,
            required: false,
            default: null
        },

        disabled: {
            required: false,
            default: false
        },

        /**
         * @deprecated tag:v6.3.0
         */
        enableInheritance: {
            type: Boolean,
            required: false,
            default: false,
            deprecated: '6.3.0'
        },

        disableSuffix: {
            type: Boolean,
            required: false,
            default: false
        },

        grossLabel: {
            type: String,
            required: false,
            default: null
        },

        netLabel: {
            type: String,
            required: false,
            default: null
        }
    },

    watch: {
        'priceForCurrency.linked': function priceLinkedWatcher(value) {
            if (value === true) {
                this.convertGrossToNet(this.priceForCurrency.gross);
            }
        },

        'taxRate.id': function taxRateWatcher() {
            if (this.priceForCurrency.linked === true) {
                this.convertGrossToNet(this.priceForCurrency.gross);
            }
        }
    },

    computed: {
        calculatePriceApiService() {
            return Application.getContainer('factory').apiService.getByName('calculate-price');
        },

        priceForCurrency: {
            get() {
                const priceForCurrency = Object.values(this.price).find((price) => {
                    return price.currencyId === this.currency.id;
                });

                // check if price exists
                if (priceForCurrency) {
                    return priceForCurrency;
                }

                // otherwise calculate values
                return {
                    gross: this.convertPrice(this.defaultPrice.gross),
                    linked: this.defaultPrice.linked,
                    net: this.convertPrice(this.defaultPrice.net)
                };
            },
            set(newValue) {
                this.priceForCurrency.gross = newValue.gross;
                this.priceForCurrency.linked = newValue.linked;
                this.priceForCurrency.net = newValue.net;
            }
        },

        isInherited() {
            const priceForCurrency = Object.values(this.price).find((price) => {
                return price.currencyId === this.currency.id;
            });

            return !priceForCurrency;
        },

        isDisabled() {
            return this.isInherited || this.disabled;
        },

        labelGross() {
            const label = this.grossLabel ? this.grossLabel : this.$tc('global.sw-price-field.labelPriceGross');
            return this.label ? label : '';
        },

        labelNet() {
            const label = this.netLabel ? this.netLabel : this.$tc('global.sw-price-field.labelPriceNet');
            return this.label ? label : '';
        },

        grossError() {
            return this.error ? this.error.gross : null;
        },

        netError() {
            return this.error ? this.error.net : null;
        }
    },

    methods: {
        onLockSwitch() {
            if (this.isDisabled) {
                return;
            }
            this.priceForCurrency.linked = !this.priceForCurrency.linked;
            this.$emit('price-lock-change', this.priceForCurrency.linked);
            this.$emit('change', this.priceForCurrency);
        },

        onPriceGrossChange(value) {
            this.priceForCurrency.gross = value;

            if (this.priceForCurrency.linked) {
                this.$emit('price-calculate', true);
                this.onPriceGrossChangeDebounce(value);
            }
        },

        onPriceGrossChangeDebounce: utils.debounce(function onPriceGrossChange(value) {
            this.$emit('price-gross-change', value);
            this.$emit('change', this.priceForCurrency);

            this.convertGrossToNet(value);
        }, 500),

        onPriceNetChange(value) {
            this.priceForCurrency.net = value;

            if (this.priceForCurrency.linked) {
                this.$emit('price-calculate', true);
                this.onPriceNetChangeDebounce(value);
            }
        },

        onPriceNetChangeDebounce: utils.debounce(function onPriceNetChange(value) {
            this.$emit('price-net-change', value);
            this.$emit('change', this.priceForCurrency);

            this.convertNetToGross(value);
        }, 500),

        convertNetToGross(value) {
            if (!value || typeof value !== 'number') {
                this.priceForCurrency.gross = 0;
                return false;
            }
            this.$emit('price-calculate', true);

            this.requestTaxValue(value, 'net').then((res) => {
                this.priceForCurrency.gross = this.priceForCurrency.net + res;
            });
            return true;
        },

        convertGrossToNet(value) {
            if (!value || typeof value !== 'number') {
                this.priceForCurrency.net = 0;
                this.$emit('calculating', false);
                return false;
            }
            this.$emit('price-calculate', true);

            this.requestTaxValue(value, 'gross').then((res) => {
                this.priceForCurrency.net = this.priceForCurrency.gross - res;
            });
            return true;
        },

        requestTaxValue(value, outputType) {
            this.$emit('price-calculate', true);
            return new Promise((resolve) => {
                if (
                    !value ||
                    typeof value !== 'number' ||
                    !this.priceForCurrency[outputType] ||
                    !this.taxRate.id ||
                    !outputType
                ) {
                    return null;
                }

                this.calculatePriceApiService.calculatePrice({
                    taxId: this.taxRate.id,
                    currencyId: this.currency.id,
                    price: this.priceForCurrency[outputType],
                    output: outputType
                }).then(({ data }) => {
                    resolve(data.calculatedTaxes[0].tax);
                    this.$emit('price-calculate', false);
                });
                return true;
            });
        },

        convertPrice(value) {
            const calculatedPrice = value * this.currency.factor;
            const priceRounded = calculatedPrice.toFixed(this.currency.decimalPrecision);
            return Number(priceRounded);
        },

        keymonitor(event) {
            if (event.key === ',') {
                const value = event.currentTarget.value;
                event.currentTarget.value = value.replace(/.$/, '.');
            }
        }
    }
});
