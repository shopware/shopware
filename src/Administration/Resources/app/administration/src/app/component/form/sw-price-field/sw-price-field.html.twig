<!-- eslint-disable-next-line sw-deprecation-rules/no-twigjs-blocks -->
{% block sw_price_field %}
<div
    class="sw-price-field"
    :class="{
        'sw-price-field--label': label,
        'sw-price-field--compact': compact,
        'sw-price-field--disabled': isDisabled
    }"
>

    <div class="price-field-grid">

        <!-- eslint-disable-next-line sw-deprecation-rules/no-twigjs-blocks -->
        {% block sw_price_field_gross %}
        {# TODO: check selectors #}
        <sw-field
            v-model="priceForCurrency.gross"
            type="number"
            class="sw-price-field__gross"
            :allow-empty="allowEmpty"
            :placeholder="$tc('sw-product.priceForm.placeholderPriceGross')"
            :label="labelGross"
            :help-text="grossHelpText"
            :validation="validation"
            :size="compact ? 'small' : 'default'"
            :min="0"
            :digits="20"
            :error="grossError"
            :disabled="isDisabled"
            :name="grossFieldName"
            v-bind="$attrs"
            @change="onPriceGrossChange"
            @keyup="keymonitor"
            @input-change="onPriceGrossInputChange"
        >
            <template
                v-if="!disableSuffix"
                #suffix
            >
                {{ currency.symbol }}
            </template>
        </sw-field>
        {% endblock %}

        <!-- eslint-disable-next-line sw-deprecation-rules/no-twigjs-blocks -->
        {% block sw_price_field_lock_button %}
        <button
            class="sw-price-field__lock"
            :class="{
                'is--locked': priceForCurrency.linked,
                'is--disabled': isDisabled
            }"
            @click="onLockSwitch"
        >
            <sw-icon
                v-if="priceForCurrency.linked"
                key="price-field-locked-indicator"
                name="regular-lock"
                size="16"
            />
            <sw-icon
                v-else
                key="price-field-locked-indicator"
                name="regular-lock-open"
                size="16"
            />
        </button>
        {% endblock %}

        <!-- eslint-disable-next-line sw-deprecation-rules/no-twigjs-blocks -->
        {% block sw_price_field_net %}
        {# TODO: check selectors #}
        <sw-field
            v-model="priceForCurrency.net"
            type="number"
            class="sw-price-field__net"
            :allow-empty="allowEmpty"
            :placeholder="$tc('sw-product.priceForm.placeholderPriceNet')"
            :label="labelNet"
            :help-text="netHelpText"
            :validation="validation"
            :size="compact ? 'small' : 'default'"
            :min="0"
            :digits="20"
            :error="netError"
            :disabled="isInherited || disabled"
            :name="netFieldName"
            v-bind="$attrs"
            @change="onPriceNetChange"
            @keyup="keymonitor"
            @input-change="onPriceNetInputChange"
        >
            <template
                v-if="!disableSuffix"
                #suffix
            >
                {{ currency.symbol }}
            </template>
        </sw-field>
        {% endblock %}
    </div>

    <sw-container
        v-if="allowModal"
        justify="end"
        class="sw-price-form__links"
    >
        <!-- eslint-disable-next-line vuejs-accessibility/click-events-have-key-events -->
        <a
            class="sw-card__quick-link"
            @click.prevent="showModal = true"
        >
            {{ $tc('sw-product.detailBase.linkMaintainCurrencyPrices') }}
            <sw-icon
                name="regular-long-arrow-right"
                small
            />
        </a>
    </sw-container>

    <sw-maintain-currencies-modal
        v-if="showModal"
        :prices="price"
        :default-price="priceForCurrency"
        :tax-rate="taxRate"
        :disabled="false"
        :hide-list-prices="hideListPrices"
        @modal-close="onCloseModal"
    />

</div>
{% endblock %}
