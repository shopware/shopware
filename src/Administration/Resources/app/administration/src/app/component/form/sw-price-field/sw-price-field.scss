@import "~scss/variables";

$sw-price-field-lock-color-default: $color-darkgray-200;
$sw-price-field-lock-color-locked: $color-shopware-brand-500;
$sw-price-field-suffix-font-size: 20px;
$sw-price-field-suffix-font-weight: 400;

.sw-price-field {
    .price-field-grid {
        display: grid;
        grid-template-columns: 1fr 30px 1fr;
        min-width: 120px;
    }

    .sw-price-field__lock {
        height: 45px;
        background: none;
        border: none;
        color: $sw-price-field-lock-color-default;
        outline: none;
        cursor: pointer;

        &.is--locked {
            color: $sw-price-field-lock-color-locked;
        }

        &.is--disabled {
            color: $color-gray-700;
            cursor: default;
        }

        .sw-icon {
            // @deprecated tag:v6.5.0 - The !important attribute can be removed
            padding: 0 3px !important;
        }
    }

    .sw-field {
        .sw-field__addition {
            font-weight: $sw-price-field-suffix-font-weight;
            font-size: $sw-price-field-suffix-font-size;
        }
    }

    &.sw-price-field--label .sw-price-field__lock {
        margin-top: 24px;
    }

    &.sw-price-field--compact .sw-price-field__lock {
        height: 32px;
    }

    &.sw-price-field--disabled {
        pointer-events: none;
    }

    .sw-price-form__links {
        grid-area: center;
    }
}
