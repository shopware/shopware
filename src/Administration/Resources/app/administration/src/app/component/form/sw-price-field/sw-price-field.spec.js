/**
 * @package admin
 */

import { shallowMount } from '@vue/test-utils';
import 'src/app/component/form/sw-price-field';

// mock data
const dollarPrice = {
    currencyId: 'a435755c6c4f4fb4b81ec32b4c07e06e',
    net: 250,
    gross: 123,
    linked: false
};

const euroPrice = {
    currencyId: 'b7d2554b0ce847cd82f3ac9bd1c0dfca',
    net: 152.33644859813083,
    gross: 163,
    linked: true
};

const taxRate = {
    name: '7%',
    taxRate: 7,
    id: 'd9eac12a83984df59a618a5be1342009'
};

const currency = {
    id: 'a435755c6c4f4fb4b81ec32b4c07e06e',
    name: 'US-Dollar',
    isoCode: 'USD',
    decimalPrecision: 2,
    factor: 1.17085,
    shortName: 'USD',
    symbol: '$'
};

const defaultPrice = {
    currencyId: 'b7d2554b0ce847cd82f3ac9bd1c0dfca',
    gross: 163,
    net: 152.33644859813083,
    linked: true
};

// initial component setup
const setup = async (propOverride) => {
    const propsData = {
        price: [dollarPrice, euroPrice],
        taxRate,
        currency,
        defaultPrice,
        enableInheritance: false,
        ...propOverride
    };

    return shallowMount(await Shopware.Component.build('sw-price-field'), {
        stubs: ['sw-field', 'sw-icon'],
        propsData
    });
};

describe('components/form/sw-price-field', () => {
    beforeEach(() => {
        Shopware.Application.getContainer = () => {
            return {
                apiService: {
                    getByName() {
                        return {
                            calculatePrice() {
                                return Promise.resolve({
                                    data: {
                                        calculatedTaxes: [],
                                    }
                                });
                            }
                        };
                    }
                }
            };
        };
    });

    it('should be a Vue.js component', async () => {
        const wrapper = await setup();
        expect(wrapper.vm).toBeTruthy();
    });

    it('should render correctly', async () => {
        const wrapper = await setup();
        expect(wrapper.element).toMatchSnapshot();
    });

    it('should contain the dollar price', async () => {
        const wrapper = await setup();
        expect(wrapper.vm.priceForCurrency.gross).toEqual(dollarPrice.gross);
        expect(wrapper.vm.priceForCurrency.net).toEqual(dollarPrice.net);
    });

    it('should not be an disabled field', async () => {
        const wrapper = await setup();
        expect(wrapper.find('.sw-price-field--disabled').exists()).toBeFalsy();
    });

    it('should be an disabled field', async () => {
        const wrapper = await setup({ price: [euroPrice] });
        expect(wrapper.find('.sw-price-field--disabled').exists()).toBeTruthy();
    });

    it('should calculate price based on default price', async () => {
        const wrapper = await setup({ price: [euroPrice] });
        const dollarPriceConverted = {
            gross: euroPrice.gross * currency.factor,
            net: euroPrice.net * currency.factor
        };

        expect(wrapper.vm.priceForCurrency.gross).toEqual(dollarPriceConverted.gross);
        expect(wrapper.vm.priceForCurrency.net + 0.0).toEqual(dollarPriceConverted.net);
    });

    it('should remove the inheritance when matching currency price exists', async () => {
        const wrapper = await setup({ price: [euroPrice] });
        expect(wrapper.vm.isInherited).toBeTruthy();
        await wrapper.setProps({ price: [dollarPrice, euroPrice] });
        expect(wrapper.vm.isInherited).toBeFalsy();
    });

    it('should set gross value null when the net value is not a number and allow empty is true', async () => {
        const wrapper = await setup({ allowEmpty: true });
        wrapper.vm.convertNetToGross(parseFloat(''));
        expect(wrapper.vm.priceForCurrency.gross).toBe(null);
    });

    it('should set gross value 0 when the net value is not a number and allow empty is false', async () => {
        const wrapper = await setup({ allowEmpty: false });
        wrapper.vm.convertNetToGross(parseFloat(''));
        expect(wrapper.vm.priceForCurrency.gross).toBe(0);
    });

    it('should set net value null when the gross value is not a number and allow empty is true', async () => {
        const wrapper = await setup({ allowEmpty: true });
        wrapper.vm.convertGrossToNet(parseFloat(''));
        expect(wrapper.vm.priceForCurrency.net).toBe(null);
    });

    it('should set net value 0 when the gross value is not a number and allow empty is false', async () => {
        const wrapper = await setup({ allowEmpty: false });
        wrapper.vm.convertGrossToNet(parseFloat(''));
        expect(wrapper.vm.priceForCurrency.net).toBe(0);
    });

    it('should calculate values if inherited and price is not set', async () => {
        const wrapper = await setup({ allowEmpty: false });
        await wrapper.setProps({
            price: [euroPrice]
        });

        const expectedNetPrice = (euroPrice.net * currency.factor);

        expect(wrapper.vm.priceForCurrency.net).toBe(parseFloat(expectedNetPrice, 10));
    });

    it('should set values to null if not inherited and price is not set', async () => {
        const wrapper = await setup({ allowEmpty: false });
        await wrapper.setProps({
            price: [euroPrice],
            inherited: false
        });

        expect(wrapper.vm.priceForCurrency.net).toBeNull();
    });

    it('should pass down gross and net helptext', async () => {
        const wrapper = await setup({
            grossHelpText: 'help for gross price',
            netHelpText: 'help for net price'
        });

        expect(wrapper.find('.sw-price-field__gross').attributes()['help-text']).toBe('help for gross price');
        expect(wrapper.find('.sw-price-field__net').attributes()['help-text']).toBe('help for net price');
    });
});
