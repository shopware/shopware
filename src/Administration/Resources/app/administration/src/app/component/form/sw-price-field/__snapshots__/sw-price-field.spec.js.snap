// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/form/sw-price-field should render correctly 1`] = `
<div
  class="sw-price-field sw-price-field--label"
>
  <div
    class="price-field-grid"
  >
    <sw-field-stub
      class="sw-price-field__gross"
      digits="20"
      label="global.sw-price-field.labelPriceGross"
      min="0"
      name="sw-price-field-gross"
      placeholder="sw-product.priceForm.placeholderPriceGross"
      size="default"
      type="number"
      value="123"
    />
     
    <button
      class="sw-price-field__lock"
    >
      <sw-icon-stub
        name="regular-lock-open"
        size="16"
      />
    </button>
     
    <sw-field-stub
      class="sw-price-field__net"
      digits="20"
      label="global.sw-price-field.labelPriceNet"
      min="0"
      name="sw-price-field-net"
      placeholder="sw-product.priceForm.placeholderPriceNet"
      size="default"
      type="number"
      value="250"
    />
  </div>
   
  <!---->
   
  <!---->
</div>
`;
